97.md

```diff
@@ -0,0 +1,251 @@
+NIP-97
+=======
+
+Attachments
+-----------
+
+`final` `optional` `author:ondra-novak`
+
+This NIP solves the problem of sharing and distributing binary content over NOSTR networks such as media - images, short videos music, or in general any files up to a certain size
+
+Key Takeaways
+-------------
+
+* Defines how binary content (images, media, files) is managed by NOSTR network
+* Allows sending private binary content in direct messages (encrypted)
+* Defines new commands at the protocol level ("ATTACH","FETCH")
+* Defines a new tag "attachment"
+* Uses the binary messages in the websocket connection
+
+
+New tag "attachment"
+-------------
+The proposal introduces a new tag  "attachment" with the following format
+
+```
+["attachment","hash_hex","size_in_bytes","mime_type","features1", "feature2", ....]
+```
+
+* **hash_hex** - hexadecimal representation of SHA256 hash of the attached binary file. This string also serves as **attachment-id**
+* **size_in_bytes** - size of the attachment in bytes
+* **mime_type** - mime type of the attachment 
+* **features...** - optional fields with various features in format `<field>=<value>`
+    * **blurhash=** - contains NIP-94 blurhash
+    * **purpose=** - purpose of the attachment in context of current note. This can help client
+to select optimal rendering. Suggested purposes
+        * **thumbnail** - thubnail image of a following video
+        * **story** - all attachments marked by this purpose are played as a story
+        * **inline** - attachment is rendered in text, referenced by `#[<index>]`. This purpose hides this attachment from the list of attachments in note's detail.
+
+```
+["attachment","12887...","64000","image/jpeg","blurhash=qwdjq3...", "purpose=thumbnail"]
+["attachment","874788...","2258147","video/mpeg4", "purpose=story"]
+["attachment","ae3c758...","2158799","video/mpeg4", "purpose=story"]
+["attachment","788eq78...","254123","image/jpeg", "purpose=story"]
+```
+        
+
+* One event can have more than one of these tags, then it carries more attachments.
```

Comment:
Using NIP-94 you would have just one event per attachment, which enables more interesting features like defining owners, liking, zapping, and reporting each attachment individually.

Reply:
So in summary, you suggest one binary content to one event. You will need to define kind for such event. Similar to NIP-95
This is also addressed in FAQ section.

This will need to two indirections which can complicate some things. How do you implement stories (very popular facebook stories). How do you manage replaceable events. I want to have nice icon and background image in "Kind: 0" Who is responsible to erase suich unused events with  attachments? Will your client implement a gallery or a file manager?

Reply:
NIP-95 can't be used for Kind:4

Reply:
> you suggest one binary content to one event. You will need to define kind for such event.

NIP-94 IS that event. It already exists, it's being used and can easily do what you need here. There is no need to define a new one. 

> This will need to two indirections which can complicate some things.

Zaps have 1-2 indirections. The New DMs will have 4. Some group chats will have 5. NIP-94 also has redirections. It's really not a big deal. 

> Will your client implement a gallery or a file manager?

We already do. There is an online client as well: https://filestr.vercel.app/ 

> Who is responsible to erase suich unused events with attachments? 

Garbage collectors already work for Zaps that don't have an event, or likes that don't have an event. This shouldn't be a big problem. Orphan events are EXTREMELY common in Nostr. 

> NIP-95 can't be used for Kind:4

They allow encryption, so I already used them like that. Same for NIP-94. I don't know why you would say it can't be used.

Reply:
OK I can accept this, you suggest binary version of NIP-95 (in summary, thank for feedback)

Reply:
>Garbage collectors already work for Zaps that don't have an event, or likes that don't have an event. This shouldn't be a big problem. Orphan events are EXTREMELY common in Nostr and some relays already deal with them.

Which NIP defines garbage collection of orphan of events?

Reply:
> Which NIP defines garbage collection of orphan of events?

I don't think there is an agreed NIP for that and we might actually not need an agreement on how GCs are run. For clients, GC doesn't really matter. It's purely a relay software/operator choice. I assume most of them just do it to reduce costs (memory and database sizes). 

Paid relays usually have tons of orphans because a paid user can cite an event that was signed by an unpaid user. Since the "reply" event author has paid, they don't delete those events. Thus, orphans are everywhere.